test/blackbox-tests/test-cases/runtest-cmd-tests.t

```diff
@@ -0,0 +1,215 @@
+  > let x = 42
+  > EOF
+
+  $ cat > dune <<EOF
+  > (library
+  >  (name mylib)
+  >  (modules lib))
+  > (tests
+  >  (names test1 test2)
+  >  (modules test1 test2 test_lib))
+  > EOF
+
+  $ dune test lib.ml
+  Error: "lib.ml" does not match any known test.
+  [1]
+
+Error when specifying a non-existent ML file:
+
+  $ dune test nonexistent.ml
+  Error: "nonexistent.ml" does not match any known test.
+  [1]
+
+  $ dune test test3.ml
+  Error: "test3.ml" does not match any known test.
+  Hint: did you mean test1.ml or test2.ml?
+  [1]
+
+Can run tests from _build directory:
+
+  $ rm -rf _build
+  $ dune test _build/default/test1.ml
+  Running test1
+
+Can specify multiple ML files:
+
+  $ rm -rf _build
+  $ dune test test1.ml test2.ml
+  Running test1
+  File "dune", line 5, characters 14-19:
+  5 |  (names test1 test2)
+                    ^^^^^
+  Running test2
+  Fatal error: exception Assert_failure("test2.ml", 3, 2)
+  [1]
+
+--------------------------------------------------------------------------------
+
+Test with a single test executable and multiple modules.
+
+  $ cat > dune <<EOF
+  > (test
+  >  (name single_test)
+  >  (modules single_test helper1 helper2))
+  > EOF
+
+  $ cat > helper1.ml <<EOF
+  > let value = "from helper1"
+  > EOF
+
+  $ cat > helper2.ml <<EOF
+  > let value = Helper1.value ^ " and helper2"
+  > EOF
+
+  $ cat > single_test.ml <<EOF
+  > let () =
+  >   Printf.printf "Test: %s\n" Helper2.value;
+  >   assert false
+  > EOF
+
+Targeting the entry point:
+
+  $ dune test single_test.ml
+  File "dune", line 2, characters 7-18:
+  2 |  (name single_test)
+             ^^^^^^^^^^^
+  Test: from helper1 and helper2
+  Fatal error: exception Assert_failure("single_test.ml", 3, 2)
+  [1]
+
+Targeting helper modules should also run the test:
+
+  $ dune test helper1.ml
+  File "dune", line 2, characters 7-18:
+  2 |  (name single_test)
+             ^^^^^^^^^^^
+  Test: from helper1 and helper2
+  Fatal error: exception Assert_failure("single_test.ml", 3, 2)
+  [1]
+
+  $ dune test helper2.ml
+  File "dune", line 2, characters 7-18:
+  2 |  (name single_test)
+             ^^^^^^^^^^^
+  Test: from helper1 and helper2
+  Fatal error: exception Assert_failure("single_test.ml", 3, 2)
+  [1]
+
+--------------------------------------------------------------------------------
+
+Test with more complex module dependencies.
+Set up: A <- B <- C -> D -> E where A and E are entry points.
+
+  $ cat > dune <<EOF
+  > (tests
+  >  (names a e)
+  >  (modules a b c d e))
+  > EOF
+
+  $ cat > c.ml <<EOF
+  > let shared = "shared by both"
+  > EOF
+
+  $ cat > b.ml <<EOF
+  > let helper = "used only by A via " ^ C.shared
+  > EOF
+
+  $ cat > d.ml <<EOF
+  > let helper = "used only by E via " ^ C.shared
+  > EOF
+
+  $ cat > a.ml <<EOF
+  > let () =
+  >   Printf.printf "A using %s\n" B.helper;
+  >   assert false
+  > EOF
+
+  $ cat > e.ml <<EOF
+  > let () =
+  >   Printf.printf "E using %s\n" D.helper;
+  >   assert false
```

Comment:
Why not just use `exit 1`? That would also make our tests less sensitive to the version of the compiler. We'd rather not rely on how the compiler pretty printing assertion failure exception